test/integration_test.go

```diff
@@ -604,6 +604,41 @@ func (ts *IntegrationTestSuite) TestHeartbeatOnActivityFailure() {
 	ts.Equal(6, heartbeatCounts)
 }
 
+func (ts *IntegrationTestSuite) TestActivityPause() {
+	ctx, cancel := context.WithTimeout(context.Background(), ctxTimeout)
+	defer cancel()
+	// Run ActivityHeartbeat workflow, this workflow will call
+	// ActivityToBePaused activity twice, the first call will test pausing an activity successfully
+	// and the second call will test completing the activity after it is resumed
+	run, err := ts.client.ExecuteWorkflow(ctx,
+		ts.startWorkflowOptions("test-activity-pause"), ts.workflows.ActivityHeartbeat)
+	ts.NoError(err)
+	// Wait for the workflow to finish
+	var result string
+	err = run.Get(ctx, &result)
+	ts.NoError(err)
+	// Check the result
+	ts.Equal("I am stopped by Pause", result)
+	// Verify that the activity was called twice
+	expectedActivities := []string{"ActivityToBePaused", "ActivityToBePaused"}
+	ts.EqualValues(expectedActivities, ts.activities.invoked())
+	// Describe the workflow execution
+	desc, err := ts.client.WorkflowService().DescribeWorkflowExecution(ctx, &workflowservice.DescribeWorkflowExecutionRequest{
+		Namespace: ts.config.Namespace,
+		Execution: &commonpb.WorkflowExecution{
+			WorkflowId: run.GetID(),
+			RunId:      run.GetRunID(),
+		},
+	})
+	ts.NoError(err)
+	// Check the workflow still has one paused pending activity
+	ts.Len(desc.GetPendingActivities(), 1)
+	ts.Equal(desc.GetPendingActivities()[0].GetActivityType().GetName(), "ActivityToBePaused")
+	ts.Equal(desc.GetPendingActivities()[0].GetAttempt(), int32(1))
+	ts.Nil(desc.GetPendingActivities()[0].GetLastFailure()) // Verify that the activity paused successfully
```

Comment:
Hrmm, it has been asked but not decided yet what this looks like. I think we should be a bit more detailed on the assertion here of what the activity failure of a pause interruption looks like. Is it a canceled failure or an application failure, and if the latter, if at all possible to have SDKs match their "type" strings here that'd be ideal.

Reply:
The API described how the SDK should respond to a failure, I don't see how this assertion can be more detailed

Reply:
How can I check the type of a nil error?

Reply:
Oh, sorry, I misread that as `NotNil`. I am not sure this is right. Hrmm, I am not sure this is correct behavior to have a nil last failure. Do last heartbeat details at least get preserved? Users expect any attempt > 1 to have a last failure, but this should an attempt could be something like 5, get paused, and have no failure.

Can we get confirmation this is the behavior we want?

Reply:
Ok, confirmed internally that we don't expect last_failure to be null. Will put server GH issue link in here once available. Want to leave assertion as is (to break on server upgrade) or want to comment out assertion with this link (when available)?

Reply:
Note, this is fixed in https://github.com/temporalio/temporal/pull/7572 and once released/used, tests relying on last failure to be nil will start failing.